src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -149,25 +149,34 @@ private Task<Releaser> LockWaitingHelper(Task waitTask)
         {
             Requires.NotNull(waitTask, nameof(waitTask));
 
-            return waitTask.IsCompleted
-                ? this.uncontestedReleaser // uncontested lock
-                : waitTask.ContinueWith(
-                    (waiter, state) =>
-                    {
-                        // Re-throw any cancellation or fault exceptions.
-                        waiter.GetAwaiter().GetResult();
-
-                        if (this.disposed)
+            switch (waitTask.Status)
+            {
+                case TaskStatus.RanToCompletion:
+                    // uncontested lock
+                    return this.uncontestedReleaser;
+
+                case TaskStatus.Canceled:
+                    return this.canceledReleaser;
+
+                default:
+                    return waitTask.ContinueWith(
+                        (waiter, state) =>
                         {
-                            throw new ObjectDisposedException(this.GetType().FullName);
-                        }
-
-                        return new Releaser((AsyncSemaphore)state);
-                    },
-                    this,
-                    CancellationToken.None,
-                    TaskContinuationOptions.ExecuteSynchronously,
-                    TaskScheduler.Default);
+                            // Re-throw any cancellation or fault exceptions.
+                            waiter.GetAwaiter().GetResult();
+
+                            if (this.disposed)
+                            {
+                                throw new ObjectDisposedException(this.GetType().FullName);
+                            }
+
+                            return new Releaser((AsyncSemaphore)state);
+                        },
+                        this,
+                        CancellationToken.None,
+                        TaskContinuationOptions.ExecuteSynchronously,
+                        TaskScheduler.Default);
+            }
```

Comment:
Alternatively, this could just be an async method:
```C#
private async Task<Releaser> LockWaitingHelper(Task waitTask)
{
    await waitTask.ConfigureAwait(false);
    if (this.disposed) throw new ObjectDisposedException(GetType().FullName);
    return new Releaser(this);
}
```

Reply:
True. We were trying to reduce allocations in the uncontested case by returning a saved task.

Reply:
Fair enough.

Reply:
I agree that the current form is visually suboptimal, especially for the case of an exception thrown in the continuation (if even possible). However, I was trying to keep the fix as close to the current code flow as possible since we're taking it through servicing.

Reply:
> I was trying to keep the fix as close to the current code flow as possible

If your goal is to keep the change as minimal as possible, wouldn't it just be a one-line change from:
```C#
return waitTask.IsCompleted
```
to:
```C#
return waitTask.Status == TaskStatus.RanToCompletion
```
?

Reply:
I believe it would leave the returned task in a faulted state instead of a cancelled state. I would be happy with that change though since technically it's already possible today.

Reply:
> I believe it would leave the returned task in a faulted state instead of a cancelled state

It would, but as you say, that's not the issue that's being fixed here.

Reply:
And the faulted task still throws the same exception. So that seems like a good way to minimize the diff.

Reply:
Now updated to the simpler form 👍